include/mata/nfa/nfa.hh

```diff
@@ -482,6 +508,8 @@ Nfa uni(const Nfa &lhs, const Nfa &rhs);
  * @param[out] prod_map Mapping of pairs of the original states (lhs_state, rhs_state) to new product states.
  * @return NFA as a product of NFAs @p lhs and @p rhs with ε-transitions preserved.
  */
+
+//preserve_epsilon optional too? Do we need both options?
```

Comment:
We talked through already, right? I would remove the parameter and preserve epsilon always. If I want to remove them, I will call e-removal afterwards. I am assuming that it would be the same computation.

Reply:
It is the same computation, yes.

Reply:
`intersection` should not introduce some new symbols (not in `nfa1` or `nfa2`), no?

Reply:
I think so, it uses epsilons when they appear on the input, the same ones, right?

Reply:
> intersection should not introduce some new symbols (not in nfa1 or nfa2), no?

Our implementation uses epsilon intersection construction. So it would introduce new symbols if we introduced an epsilon symbol (say a symbol with a value of 10 when our largest used symbol in the automaton is 9) each time when calling intersection, or concatenation, etc., with the preserving of epsilons turned off by default. You would get exactly what I describe: `intersection(nfa1, nfa2)` would result with:
```
Nfa {
    // ...
   0 -(4)-> 5,
   3 -(8)-> 2,
   1 -(9)-> 7,
   4 -(10)-> 3,
   1 -(10)-> 6,
}
```
How can the user know what the symbol `10` means? We agreed that the default epsilon of `Symbol::max_value()` should be omnipresent, and we could always use these for such operations, but again, operations not counting with epsilon symbols would then use the result and parse `Symbol::max_value()` as a regular symbol (as per our agreement). We would need to add a check for this specific epsilon symbol to each operation in the entire Mata library. I think the default should be not to introduce new symbols by any operation. But we could turn keeping them on explicitly, if we want, by the passed optional parameter.

We can set the default to preserving epsilon transitions, but that would mean that 90 % of the code calling intersection would look like the following:
```cpp
Nfa result = intersection(nfa1, nfa2).remove_epsilon(newly_introduced_epsilon_probably_the_default_epsilon_which_i_have_to_somehow_extract_from_the_intersection_function_call);
```
If this is the desired output and we get used to it, that yes, we can remove the optional parameter and keep epsilons always.

I think that what we talked about was removing the optional parameter for trim to reduce the result afterwards, forcing the user to manually type `nfa.trim().reduce()`, which I find reasonable with chaining functionality implemented.

Reply:
I also agree that we should always preserve epsilons. If neither of the automata contain epsilons, then the result will not contain epsilons either, so you do not have to always call `remove_epsilon`.

Reply:
now it is done that way